File: SpaceDock/endpoints/access.py
from flask import request
from sqlalchemy import desc
from SpaceDock.common import get_param, user_has, with_session
from SpaceDock.database import db
from SpaceDock.formatting import  ability_format, bulk
from SpaceDock.objects import Ability, Role, User
from SpaceDock.routing import route

import json

@route('/api/access')
@user_has('access-view')
def roles_list():
    """
    Displays  list of all roles with the matching abilities
    """
    roles = list()
    for role in Role.query.order_by(desc(Role.id)).all():
        roles.append({'id': role.id, 'name': role.name, 'abilities': bulk(role.abilities, ability_format), 'params': json.loads(role.params)})
    return {'error': False, 'count': len(roles), 'data': roles}

@route('/api/access/roles/assign', methods=['POST'])
@user_has('access-edit')
@with_session
def add_role():
    """
    Promotes a user for the given role. Required parameters: userid, rolename
    """
    userid = request.json.get('userid')
    rolename = request.json.get('rolename')
    if not isinstance(userid, int) or not User.query.filter(User.id == userid).first():
        return {'error': True, 'reasons': ['The userid is invalid.'], 'codes': ['2145']}, 400
    user = User.query.filter(User.id == userid).first()
    user.add_roles(rolename)
    db.add(Role.query.filter(Role.name == rolename).first())
    return {'error': False}

@route('/api/access/roles/remove', methods=['POST'])
@user_has('access-edit')
@with_session
def remove_role():
    """
    Removes a user from a group. Required parameters: userid, rolename
    """
    userid = request.json.get('userid')
    rolename = request.json.get('rolename')
    if not isinstance(userid, int) or not User.query.filter(User.id == userid).first():
        return {'error': True, 'reasons': ['The userid is invalid.'], 'codes': ['2145']}, 400
    user = User.query.filter(User.id == userid).first()
    if not rolename in user._roles:
        return {'error': True, 'reasons': ['The user doesn\'t have this role'], 'codes': ['1015']}, 400
    user.remove_roles(rolename)
    db.delete(Role.query.filter(Role.name == rolename).first())
    return {'error': False}

@route('/api/access/abilities')
@user_has('access-view')
def abilities_list():
    """
    Displays a list of all abilities.
    """
    abilities = list()
    for ability in Ability.query.order_by(desc(Ability.id)).all():
        abilities.append(ability_format(ability))
    return {'error': False, 'count': len(abilities), 'data': abilities}

@route('/api/access/abilities/assign', methods=['POST'])
@user_has('access-edit')
@with_session
def add_abilities():
    """
    Adds a permission to a group. Required parameters: rolename, abname
    """
    rolename = request.json.get('rolename')
    abname = request.json.get('abname')
    if not Role.query.filter(Role.name == rolename).first():
        return {'error': True, 'reasons': ['The role does not exist. Please add it to a user to create it internally.'], 'codes': ['3030']}, 400
    r = Role.query.filter(Role.name == rolename).first()
    r.add_abilities(abname)
    return {'error': False}

@route('/api/access/abilities/remove', methods=['POST'])
@user_has('access-edit')
@with_session
def remove_abilities():
    """
    Removes a permission from a group. Required parameters: rolename, abname
    """
    rolename = request.json.get('rolename')
    abname = request.json.get('abname')
    errors = ()
    codes = ()
    if not Role.query.filter(Role.name == rolename).first():
        errors.append('The role does not exist.')
        codes.append('3030')
    if not Ability.query.filter(Ability.name == abname).first():
        errors.append('The ability does not exist.')
        codes.append('2107')
    if len(errors) > 0:
        return {'error': True, 'reasons': errors, 'codes' codes}, 400
    role = Role.query.filter(Role.name == rolename).first()
    ability = Ability.query.filter(Ability.name == abname).first()
    if not ability in role.abilities:
        return {'error': True, 'reasons': ['The ability isn\'t assigned to this role'], 'codes': ['1010']}, 400
    role.remove_abilities(abname)
    return {'error': False}

@route('/api/access/params/add/<rolename>', methods=['POST'])
@user_has('access-edit')
@with_session
def add_params(rolename):
    """
    Adds a parameter for an ability. Required parameters: abname, param, value
    """
    abname = request.json.get('abname')
    param = request.json.get('param')
    value = request.json.get('value')
    errors = ()
    codes = ()
    if not Role.query.filter(Role.name == rolename).first():
        errors.append('The rolename is invalid.')
        codes.append('3030')
    if not Ability.query.filter(Ability.name == abname).first():
        errors.append('The ability does not exist.')
        codes.append('2107')
    if len(errors) > 0:
        return {'error': True, 'reasons': errors, 'codes': codes}, 400
    role = Role.query.filter(Role.name == rolename).first()
    role.add_param(abname, param, value)
    return {'error': False}

@route('/api/access/params/remove/<rolename>', methods=['POST'])
@user_has('access-edit')
@with_session
def remove_params(rolename):
    """
    Removes a parameter from an ability. Required parameters: abname, param, value
    """
    abname = request.json.get('abname')
    param = request.json.get('param')
    value = request.json.get('value')
    errors = ()
    codes = ()
    if not Role.query.filter(Role.name == rolename).first():
        errors.append('The rolename is invalid.')
        codes.append('3030')
    if not Ability.query.filter(Ability.name == abname).first():
        errors.append('The ability does not exist.')
        codes.append('2107')
    if len(errors) > 0:
        return {'error': True, 'reasons': errors, 'codes': codes}, 400
    role = Role.query.filter(Role.name == rolename).first()
    if not value in get_param(abname, param, json.loads(role.params)):
        return {'error': True, 'reasons': ['The parameter doesn\'t exist'], 'codes': ['2140']}, 400
    role.remove_param(abname, param, value)
    return {'error': False}

File: src/SpaceDock/routes/access.go
/*
 SpaceDock Backend
 API Backend for the SpaceDock infrastructure to host modfiles for various games

 SpaceDock Backend is licensed under the Terms of the MIT License.
 Copyright (c) 2017 Dorian Stoll (StollD), RockyTV
*/

package routes

import (
    "SpaceDock"
    "SpaceDock/middleware"
    "SpaceDock/objects"
    "SpaceDock/utils"
    "github.com/spf13/cast"
    "gopkg.in/kataras/iris.v6"
)

/*
 Registers the routes for the account management
 */
func AccessRegister() {
    Register(GET, "/api/access/roles",
        middleware.NeedsPermission("access-view", true),
        list_roles,
    )
    Register(POST, "/api/access/roles",
        middleware.NeedsPermission("access-edit", true),
        assign_role,
    )
    Register(DELETE, "/api/access/roles",
        middleware.NeedsPermission("access-edit", true),
        remove_role,
    )
    Register(GET, "/api/access/abilities",
        middleware.NeedsPermission("access-view", true),
        list_abilities,
    )
    Register(POST, "/api/access/abilities",
        middleware.NeedsPermission("access-edit", true),
        assign_ability,
    )
    Register(DELETE, "/api/access/abilities",
        middleware.NeedsPermission("access-edit", true),
        remove_ability,
    )
    Register(POST, "/api/access/roles/:rolename/params",
        middleware.NeedsPermission("access-edit", true),
        add_param,
    )
    Register(DELETE, "/api/access/roles/:rolename/params",
        middleware.NeedsPermission("access-edit", true),
        remove_param,
    )
}

/*
 Path: /api/access/roles/
 Method: GET
 Description: Displays  list of all roles with the matching abilities
 Abilities: access-view
 */
func list_roles(ctx *iris.Context) {
    roles := []objects.Role{}
    SpaceDock.Database.Find(&roles)
    output := make([]map[string]interface{}, len(roles))
    for i,element := range roles {
        abilities := element.Abilities
        abilitiesout := make([]map[string]interface{}, len(abilities))
        for j,element2 := range abilities {
            abilitiesout[j] = map[string]interface{} {
                "id": element2.ID,
                "name": element2.Name,
                "meta": utils.LoadJSON(element2.Meta),
            }
        }
        output[i] = map[string]interface{} {
            "id": element.ID,
            "name": element.Name,
            "abilities": abilitiesout,
            "meta": utils.LoadJSON(element.Meta),
        }
    }
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false, "count": len(roles), "data": output})
}

/*
 Path: /api/access/roles/
 Method: POST
 Description: Promotes a user for the given role. Required parameters: userid, rolename
 Abilities: access-edit
 */
func assign_role(ctx *iris.Context) {
    // Grab parameters from the JSON
    userid := cast.ToUint(utils.GetJSON(ctx,"userid"))
    rolename := cast.ToString(utils.GetJSON(ctx,"rolename"))

    // Try to get the user
    user := &objects.User{}
    SpaceDock.Database.Where("id = ?", userid).First(user)
    if user.ID != userid {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("The userid is invalid.").Code(2145))
    }

    // User is valid, assign the new role
    role := user.AddRole(rolename)
    SpaceDock.Database.Save(role).Save(user)
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false})
}



/*
 Path: /api/access/roles/
 Method: DELETE
 Description: Promotes a user for the given role. Required parameters: userid, rolename
 Abilities: access-edit
 */
func remove_role(ctx *iris.Context) {
    // Grab parameters from the JSON
    userid := cast.ToUint(utils.GetJSON(ctx,"userid"))
    rolename := cast.ToString(utils.GetJSON(ctx,"rolename"))

    // Try to get the user
    user := &objects.User{}
    SpaceDock.Database.Where("id = ?", userid).First(user)
    if user.ID != userid {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("The userid is invalid.").Code(2145))
        return
    }
    if user.HasRole(rolename) {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("The user doesn't have this role").Code(1015))
        return
    }

    // Everything is valid, remove the role
    user.RemoveRole(rolename)
    SpaceDock.Database.Save(user)
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false})
}

/*
 Path: /api/access/abilities
 Method: GET
 Description: Displays a list of all abilities.
 Abilities: access-view
 */
func list_abilities(ctx *iris.Context) {
    var abilities []objects.Ability
    SpaceDock.Database.Find(&abilities)
    output := make([]map[string]interface{}, len(abilities))
    for i,element := range abilities {
        output[i] = map[string]interface{} {
            "id": element.ID,
            "name": element.Name,
            "meta": utils.LoadJSON(element.Meta),
        }
    }
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false, "count": len(abilities), "data": output})
}

/*
 Path: /api/access/abilities/
 Method: POST
 Description: Adds a permission to a group. Required parameters: rolename, abname
 Abilities: access-edit
 */
func assign_ability(ctx *iris.Context) {
    // Grab parameters from the JSON
    rolename := cast.ToString(utils.GetJSON(ctx,"rolename"))
    abname := cast.ToString(utils.GetJSON(ctx,"abname"))

    // Try to get the role
    role := &objects.Role{}
    SpaceDock.Database.Where("name = ?", rolename).First(role)
    if role.Name != rolename {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("The role does not exist. Please add it to a user to create it internally.").Code(3030))
        return
    }

    // Role is valid, assign the new ability
    ability := role.AddAbility(abname)
    SpaceDock.Database.Save(role).Save(ability)
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false})
}

/*
 Path: /api/access/abilities/
 Method: DELETE
 Description: Removes a permission from a group. Required parameters: rolename, abname
 Abilities: access-edit
 */
func remove_ability(ctx *iris.Context) {
    // Grab parameters from the JSON
    rolename := cast.ToString(utils.GetJSON(ctx,"rolename"))
    abname := cast.ToString(utils.GetJSON(ctx,"abname"))

    // Try to get the role
    role := &objects.Role{}
    SpaceDock.Database.Where("name = ?", rolename).First(role)
    ability := &objects.Ability{}
    SpaceDock.Database.Where("name = ?", abname).First(ability)
    errors := []string{}
    codes := []int{}
    if role.Name != rolename {
        errors = append(errors,"The role does not exist.")
        codes = append(codes, 3030)
    }
    if ability.Name != abname {
        errors = append(errors,"The ability does not exist.")
        codes = append(codes, 2107)
    }
    if len(errors) > 0 {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error(errors...).Code(codes...))
        return
    }

    // Both objects are valid, check if they are linked
    if !role.HasAbility(abname) {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("The ability isn't assigned to this role").Code(1010))
        return
    }

    // Remove the ability
    role.RemoveAbility(abname)
    SpaceDock.Database.Save(role)
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false})
}

/*
 Path: /api/access/roles/:rolename/params/
 Method: POST
 Description: Adds a parameter for an ability. Required parameters: abname, param, value
 Abilities: access-edit
 */
func add_param(ctx *iris.Context) {
    // Grab parameters from the JSON
    rolename := ctx.GetString("rolename")
    abname := cast.ToString(utils.GetJSON(ctx,"abname"))
    param := cast.ToString(utils.GetJSON(ctx,"param"))
    value := cast.ToString(utils.GetJSON(ctx,"value"))

    // Try to get the role
    role := &objects.Role{}
    SpaceDock.Database.Where("name = ?", rolename).First(role)
    ability := &objects.Ability{}
    SpaceDock.Database.Where("name = ?", abname).First(ability)
    errors := []string{}
    codes := []int{}
    if role.Name != rolename {
        errors = append(errors,"The role does not exist.")
        codes = append(codes, 3030)
    }
    if ability.Name != abname {
        errors = append(errors,"The ability does not exist.")
        codes = append(codes, 2107)
    }
    if len(errors) > 0 {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error(errors...).Code(codes...))
        return
    }

    // Both objects are valid, check if they are linked
    role.AddParam(abname, param, value)
    SpaceDock.Database.Save(role)
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false})
}

/*
 Path: /api/access/roles/:rolename/params/
 Method: DELETE
 Description: Removes a parameter from an ability. Required parameters: abname, param, value
 Abilities: access-edit
 */
func remove_param(ctx *iris.Context) {
    // Grab parameters from the JSON
    rolename := ctx.GetString("rolename")
    abname := cast.ToString(utils.GetJSON(ctx,"abname"))
    param := cast.ToString(utils.GetJSON(ctx,"param"))
    value := cast.ToString(utils.GetJSON(ctx,"value"))

    // Try to get the role
    role := &objects.Role{}
    SpaceDock.Database.Where("name = ?", rolename).First(role)
    ability := &objects.Ability{}
    SpaceDock.Database.Where("name = ?", abname).First(ability)
    errors := []string{}
    codes := []int{}
    if role.Name != rolename {
        errors = append(errors,"The role does not exist.")
        codes = append(codes, 3030)
    }
    if ability.Name != abname {
        errors = append(errors,"The ability does not exist.")
        codes = append(codes, 2107)
    }
    if len(errors) > 0 {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error(errors...).Code(codes...))
        return
    }

    // Both objects are valid, check if the param exists
    role.RemoveParam(abname, param, value)
    SpaceDock.Database.Save(role)
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false})
}